nat-lab/tests/test_batching.py

```diff
@@ -0,0 +1,207 @@
+                    ["stun", "upnp", "local"],
+                    direct_enabled=True,
+                    batching_enabled=True,
+                ),
+                _generate_setup_parameters(
+                    ConnectionTag.DOCKER_UPNP_CLIENT_1,
+                    AdapterType.LinuxNativeWg,
+                    ["stun", "upnp", "local"],
+                    direct_enabled=True,
+                    batching_enabled=True,
+                ),
+                _generate_setup_parameters(
+                    ConnectionTag.DOCKER_UPNP_CLIENT_2,
+                    AdapterType.LinuxNativeWg,
+                    ["stun", "upnp", "local"],
+                    direct_enabled=True,
+                    batching_enabled=True,
+                ),
+            ],
+            marks=[pytest.mark.batching],
+        ),
+    ],
+)
+async def test_batching_star(setup_params: List[SetupParameters], request) -> None:
+    async with AsyncExitStack() as exit_stack:
+        capture_duration = 30
+
+        env = await setup_mesh_nodes(exit_stack, setup_params)
+
+        alpha_client, beta_client, gamma_client, delta_client = env.clients
+        alpha, beta, gamma, delta = env.nodes
+
+        clients = [alpha_client, beta_client, gamma_client, delta_client]
+        peer_nodes = [alpha, beta, gamma, delta]
+
+        await asyncio.gather(*(
+            client.wait_for_state_peer(
+                peer_node.public_key, [State.Connected], [PathType.Direct]
+            )
+            for client, peer_node in product(clients, peer_nodes)
+            if not client.is_node(peer_node)
+        ))
+
+        # Capturing the traffic can be done on nodes or gateways
+        # On gateways there's a guarantee that the traffic has arrived
+        # however we can't decrypt it easily.
+        # On nodes there's no guarantee the packet left the network adapter
+        # but we can inspect the packet before it hit the TUN device.
+        # Here we capture on both
+        gateways = [DOCKER_GW_MAP[param.connection_tag] for param in setup_params]
+        gateway_container_names = [container_id(conn_tag) for conn_tag in gateways]
+        conns = [client.get_connection() for client in env.clients]
+        node_container_names = [
+            conn.container_name()
+            for conn in conns
+            if isinstance(conn, DockerConnection)
+        ]
+
+        container_names = gateway_container_names + node_container_names
+        # TODO: assert these are docker connections
+        print("Will capture batching on containers: ", container_names)
+        cnodes = zip(env.clients, env.nodes, [c.connection for c in env.connections])
+
+        # the plan is to misalign the peers for better batcing observation. WIthout misalignment
+        # peers will form a meshnet almost immediately, already achieving the desired batching effect.
+        for client in env.clients:
+            await client.stop_device()
+
+        tasks = []
```

Comment:
Some name indicating what these tasks are doing would be nice imho

Reply:
I renamed the variable into `pcap_capture_tasks`